### PR TITLE
Rate limiting based span drop

### PR DESCRIPTION
**Pull request by SrikarMannepalli**

### CODE REVIEW

File: span-normalizer/helm/templates/span-normalizer-config.yaml [hunk 1]
```diff
@@ -106,3 +106,14 @@ data:
         expireAfterWriteDuration = {{ .Values.spanNormalizerConfig.excludeSpanRulesConfig.cache.expireAfterWriteDuration }}
       }
     }
+    {{- if hasKey .Values.spanNormalizerConfig "rateLimitConfig" }}
+      rate.limit.config = [
+        {{- range $k,$v := $.Values.spanNormalizerConfig.rateLimitConfig }}
```
Comment (avinashkolluru) [inline span-normalizer/helm/templates/span-normalizer-config.yaml:111]: Did you try with sample values? rate.limit.config is a list of dict/object. What does a range with key/value mean?

Comment (SrikarMannepalli) [reply span-normalizer/helm/templates/span-normalizer-config.yaml:111]: Yes Avinash. In this case, the key would be the index of the object and the value would be the object. It is being rendered as follows.
```rate.limit.config = [
          {
            tenantId = tenant1
            groupingKey = servicename
            maxSpansPerMinute = 100
          },
          {
            tenantId = tenant2
            groupingKey = http.method
            maxSpansPerMinute = 1000
          },
      ]
```

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/RateLimitingSpanFilter.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      tenantSpanCountMap.put(key, Map.of(0L, 0L));
+      seenAttributes.add(groupingKey);
+    }
+  }
+
+  public boolean shouldDropSpan(String tenantId, Event event) {
+    if (seenAttributes.stream()
+        .noneMatch(
+            attribute ->
+                tenantMaxSpansPerMinuteMap.containsKey(generateKey(tenantId, attribute)))) {
```
Comment (skjindal93) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/RateLimitingSpanFilter.java:37]: Can we extract this out into a separate boolean function?

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/RateLimitingSpanFilter.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      return false;
+    }
+
+    boolean shouldDropSpan = false;
+    for (String attribute : seenAttributes) {
```
Comment (skjindal93) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/RateLimitingSpanFilter.java:42]: We are iterating over the seen attributes again. Can we instead get rid of the stream above?

Comment (SrikarMannepalli) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/RateLimitingSpanFilter.java:35]: Right, updated.

**Review by skjindal93 — APPROVED** [review]

**Pull request merged**